spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitioner.scala

```diff
@@ -0,0 +1,146 @@
+      override def end(range: ITR): T = range.range.end
+      override def isFull(range: ITR): Boolean = range.range.isFull
+      override def contains(range: ITR, token: T): Boolean = range.range.contains(token)
+    }
+  }
+
+}
+
+/**
+  * A [[org.apache.spark.Partitioner]] implementation which performs the inverse
+  * operation of a traditional C* hashing. Requires the Key type and the token
+  * value type V.
+  *
+  * Will take objects of type Key and determine given the token ranges in `indexedRanges`
+  * which range the Key would belong in given the C* schema in `TableDef`
+  *
+  * Under the hood uses a bound statement to generate routing keys which are then
+  * used the driver's internal token factories to determine the token for the
+  * routing key.
+  */
+private[connector] class CassandraPartitioner[Key : ClassTag, V, T <: Token[V]](
+  private val connector: CassandraConnector,
+  private val tableDef: TableDef,
+  val partitions: Seq[CassandraPartition[V, T]],
+  val keyMapping: ColumnSelector = PartitionKeyColumns)(
+implicit
+  @transient
+  rwf: RowWriterFactory[Key],
+  tokenFactory: TokenFactory[V, T]) extends Partitioner with Logging {
+
+  /** Changes the tableDef target of this partitioner. Can only be done within a keyspace
+    * verification of key mapping will occur with the call to [[verify()]] */
+  def withTableDef(tableDef: TableDef): CassandraPartitioner[Key, V, T] = {
+    if (tableDef.keyspaceName != this.tableDef.keyspaceName) {
+      throw new IllegalArgumentException(
+        s"""Cannot apply partitioner from keyspace
+           |${this.tableDef.keyspaceName} to table
+           |${tableDef.keyspaceName}.${tableDef.tableName} because the keyspaces do
+           |not match""".stripMargin)
+    }
+
+    new CassandraPartitioner(connector, tableDef, partitions, keyMapping)
+  }
+
+  /** Changes the current key mapping for this partitioner. Verification of the mapping
+    * occurs on call to [[verify()]] */
+  def withKeyMapping(keyMapping: ColumnSelector): CassandraPartitioner[Key, V, T] =
+    new CassandraPartitioner(connector, tableDef, partitions, keyMapping)
+
+  private lazy val partitionKeyNames =
+    PartitionKeyColumns.selectFrom(tableDef).map(_.columnName).toSet
+
+  private lazy val partitionKeyMapping = keyMapping
+    .selectFrom(tableDef)
+    .filter( colRef => partitionKeyNames.contains(colRef.columnName))
+
+  private lazy val partitionKeyWriter = {
+    logDebug(
+      s"""Building Partitioner with mapping
+         |${partitionKeyMapping.map(x => (x.columnName, x.selectedAs))}
+         |for table $tableDef""".stripMargin)
+    implicitly[RowWriterFactory[Key]]
+      .rowWriter(tableDef, partitionKeyMapping)
+  }
+
+  /** Builds and makes sure we can make a rowWriter with the current TableDef and keyMapper */
+  def verify(log: Boolean = true): Unit = {
+    val attempt = Try(partitionKeyWriter)
+    if (attempt.isFailure) {
+      if (log)
```

Comment:
Here  the issue is that we sometimes try to make a a partitioner and we expect it to fail because we have no other way of checking whether or not there is valid partitioner. For those times when we expect a failure we don't want to log, but if somehow the user is about run with a broken partitioner we need to actually break with a nice message.